---
title: "Analyzing TCGA (patient) Imputation Data"
output: html_notebook
---

To work on: 
   * need to (most likely) remove the eta.sq() results
   * need to redo the TNBC analysis, removing eta.sq() component
   * need to decide on clincal or PAM50 subtyping 


Analysis outline: 

  * Install TCGAbiolinks to get TCGA clinical data
  * Format BRCA predictions, merge with clinical data
  * Perform t-tests and correct for multiple testing 

```{r, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE)
```


```{r install packages, include= FALSE}

if (!requireNamespace("TCGAbiolinks", quietly = TRUE))
    BiocManager::install("TCGAbiolinks")

if (!requireNamespace(c("broom", "grid", "ggrepel", "lsr"), quietly = TRUE))
    install.packages("broom", "grid", "ggrepel", "lsr")


library(TCGAbiolinks)
library(tidyverse)
library(broom)
library(readxl)
library(grid)
library(ggrepel)

theme_set(theme_bw())


query <- GDCquery(project = "TCGA-BRCA", 
                  data.category = "Clinical",
                  data.type = "Clinical Supplement", 
                  data.format = "BCR Biotab")
GDCdownload(query)
clinical.BCRtab.all <- GDCprepare(query)
names(clinical.BCRtab.all)
BRCA_clinical <- clinical.BCRtab.all$clinical_patient_brca %>%
    dplyr::select(bcr_patient_barcode, er_status_by_ihc, er_status_ihc_Percent_Positive, pr_status_by_ihc, 
                  pr_status_ihc_percent_positive, her2_status_by_ihc, her2_fish_status, ) %>%
    dplyr::slice(-1:-2) %>% 
    type_convert() %>% 
    mutate(HER2_Subtype = if_else(condition = her2_fish_status %in% c("Positive", "Negative"), her2_fish_status, her2_status_by_ihc)) %>% 
    filter(er_status_by_ihc %in% c("Positive", "Negative"), 
           pr_status_by_ihc %in% c("Positive", "Negative"), 
           HER2_Subtype %in% c("Positive", "Negative")) %>% 
    mutate(Clin_Subtype = if_else(condition = er_status_by_ihc == "Positive" | pr_status_by_ihc == "Positive", 
                                  true = if_else(condition = HER2_Subtype == "Positive", true = "ER/PR+, HER2+", false = "ER/PR+"),
                                  false = if_else(condition = HER2_Subtype == "Positive", true = "HER2+", false = "TNBC"))) %>% 
    mutate(TNBC_status = if_else(Clin_Subtype == "TNBC", "TNBC", "RPBC"))

TCGA_subs <- TCGAquery_subtype(tumor = "brca") #easily get TCGA PAM50 status this way

BRCA_pam50 <- TCGA_subs %>% select(patient, PAM50 = BRCA_Subtype_PAM50) %>% filter(PAM50 != "NA")

BRCA_clinical <- left_join(BRCA_clinical, BRCA_pam50, by = c("bcr_patient_barcode" = "patient")) 

######################## note ###################
#####I could look into doing the KI-67 subtyping to futher delineate the ER/PR+ samples, but TCGA doesn't have KI-67 staining. I could potentially use expression, which is what I have below, but the cutoff wouldn't be clear (expression isn't used clinically), so I have opted to skip it for now. 
# tpmDatMat <- read.delim(file = "./DataIn/TCGA/RNA-seq/BRCA.rnaseqv2__illuminahiseq_rnaseqv2__unc_edu__Level_3__RSEM_genes__data.data.txt", as.is=T)
# tpmDatMat <- tpmDatMat[c(1,grep(pattern = "MKI67", x = tpmDatMat[,"Hybridization.REF"])),]
# tpmDatMat_tpm <- tpmDatMat[-1,which(tpmDatMat[1,] == "scaled_estimate")]
# tpmDatMat_tpm <- apply(tpmDatMat_tpm, 2, as.numeric)
# geneNames <- do.call(cbind, strsplit(tpmDatMat[, "Hybridization.REF"], "|", fixed=TRUE))[1,][-1]
# rownames(tpmDatMat_tpm) <- geneNames
# colnames(tpmDatMat_tpm) <- gsub("\\.", "-", x = substr(colnames(tpmDatMat_tpm), 1, 16))
# tpmDatMat_tpm_logged <- log((tpmDatMat_tpm*1000000)+1) # transform the data
# tpmDatMat_tpm_logged <- tpmDatMat_tpm_logged[,-grep(pattern = ".11A.|.11B.|.10A.|.10B.", colnames(tpmDatMat_tpm_logged))]
# KI67_expression <- t(tpmDatMat_tpm_logged) %>% as_tibble(rownames = "bcr_patient_barcode") %>%
#     mutate(bcr_patient_barcode = substr(bcr_patient_barcode, 1, 12))
# 
# BRCA_clin_ki67 <- left_join(BRCA_clinical, KI67_expression)
# 
# BRCA_clin_ki67 %>%
#     filter(Clin_Subtype == "ER/PR+") %>%
#     filter(MKI67 < 1.8)

```

This table shows the clinical breakdown by subtype. 

```{r}
table(BRCA_clinical$Clin_Subtype)
table(BRCA_clinical$PAM50)

```

Ki-67 staining is not available, so we cannot do the traditional Lum A and Lum B groups for the clinical subtyping.  

### Prediction Data

Here we load and format the prediction dataset, then we merge it with the clinical data we previously downloaded

```{r format predictions, include = FALSE}
load("./DataOut/2015CTRP-Models_on_TCGA_BRCA.RData")
BRCA_preds <- as_tibble(t(CTRPDrugPredictions_mat), rownames = "Patient_ID") 
BRCA_preds <- BRCA_preds %>% 
  filter(grepl(Patient_ID, pattern = "\\.01A\\.|\\.01B\\.")) %>% 
  mutate(Patient_ID = gsub(x = strtrim(Patient_ID, width = 12), pattern = "\\.", replacement = "-")) %>% 
  gather(key = "Drug", value = "PSS", -Patient_ID)

#join with clinical information
preds.subtype <- left_join(BRCA_preds, BRCA_clinical, by = c("Patient_ID" = "bcr_patient_barcode")) %>% 
  drop_na(TNBC_status, PAM50)

```

We can subset the predictions and perform t-tests. 

```{r Statistical testing}
aov_preds.subtype <- preds.subtype %>%
  group_by(Drug) %>%
  do(tidy(aov(.$PSS ~ .$PAM50))) %>%
  ungroup() %>% 
  filter(term == ".$PAM50") %>%
  mutate(bonf_p.value = p.adjust(p.value, method = "bonferroni"))

clin_aov_preds.subtype <- preds.subtype %>%
  group_by(Drug) %>%
  do(tidy(aov(.$PSS ~ .$Clin_Subtype))) %>%
  ungroup() %>% 
  filter(term == ".$Clin_Subtype") %>%
  mutate(bonf_p.value = p.adjust(p.value, method = "bonferroni"))

t.tst_preds.subtype <- preds.subtype %>% 
  group_by(Drug) %>% 
  do(tidy(t.test(PSS~TNBC_status, data = .))) %>%
  ungroup() %>% 
  mutate(bonf_p.value = p.adjust(p.value, method = "bonferroni"))

t.tst_preds.subtype %>% mutate(Sensitive = if_else(estimate > 0, "TNBC", "RPBC")) %>% count(Sensitive)

```


As a note/aside: 
Based on t-tests, even with a bonferroni correction, most of the values are statistically signifcant. Ineed `r sum(t.tst_preds.subtype$bonf_p.value < 0.05/length(unique(t.tst_preds.subtype$Drug)), na.rm = TRUE)`% are statistically significant out of the `r length(unique(t.tst_preds.subtype$Drug))`total drugs, which can be seen by this histogram of t.test results and Anova Results 

```{r, echo= FALSE}

t.tst_preds.subtype %>% 
  arrange(bonf_p.value) %>% 
  mutate(log_p.correct = log10(bonf_p.value)) %>% 
  ggplot() +
    geom_histogram(mapping = aes(x = log_p.correct)) +
    geom_vline(xintercept = log10(0.05), color = "red") +
    labs(title = "Histogram of T-test Results by Signifcance")

aov_preds.subtype %>% 
  arrange(bonf_p.value) %>% 
  mutate(log_p.correct = log10(bonf_p.value)) %>% 
  ggplot() +
    geom_histogram(mapping = aes(x = log_p.correct)) +
    geom_vline(xintercept = log10(0.05), color = "red") +
    labs(title = "Histogram of Anova Results by Signifcance")

```


Let us see if this recapitulates certain expected subtype specific response. 


```{r graphing function}

PAM50_ISS_grapher <- function(drug){
  grob_annot <- grobTree(textGrob(paste("p-value =", 
                                      format(aov_preds.subtype$bonf_p.value[aov_preds.subtype$Drug == drug], 
                                             trim = T, digits = 4)), 
                                x = 0.73, y = 0.95, hjust = 0, gp = gpar(fontface = "italic")))
  print(preds.subtype %>% 
    filter(Drug == drug) %>% 
    ggplot(aes(x = PAM50, y = PSS, fill = PAM50)) + 
      geom_boxplot() +
      labs(title = paste(drug, "'s Imputed Sensitivity in TCGA BRCA Cohort by PAM50 Subtype", sep = ""),
           y = "Imputed Sensitivity Score") +
      theme(legend.position = "none",
            plot.title = element_text(hjust = 0.5))+
      annotation_custom(grob_annot))
      
}

Clin.sub_ISS_grapher <- function(drug){
  grob_annot <- grobTree(textGrob(paste("p-value =", 
                                      format(clin_aov_preds.subtype$bonf_p.value[clin_aov_preds.subtype$Drug == drug], 
                                             trim = T, digits = 4)), 
                                x = 0.73, y = 0.95, hjust = 0, gp = gpar(fontface = "italic")))
  print(preds.subtype %>% 
    filter(Drug == drug) %>% 
    ggplot(aes(x = Clin_Subtype, y = PSS, fill = Clin_Subtype)) + 
      geom_boxplot() +
      labs(title = paste(drug, "'s Imputed Sensitivity in TCGA BRCA Cohort by Clinical Subtype", sep = ""),
           y = "Imputed Sensitivity Score") +
      theme(legend.position = "none",
            plot.title = element_text(hjust = 0.5))+
      annotation_custom(grob_annot))
      
}

TNBC_ISS_grapher <- function(drug){
  grob_annot <- grobTree(textGrob(paste("p-value =", 
                                      format(t.tst_preds.subtype$bonf_p.value[t.tst_preds.subtype$Drug == drug], 
                                             trim = T, digits = 4)), 
                                x = 0.73, y = 0.95, hjust = 0, gp = gpar(fontface = "italic")))
  print(preds.subtype %>% 
    filter(Drug == drug) %>% 
    ggplot(aes(x = TNBC_status, y = PSS, fill = TNBC_status)) + 
      geom_boxplot() +
      labs(title = paste(drug, "'s Imputed Sensitivity in TCGA BRCA Cohort by TNBC Status", sep = ""),
           y = "Imputed Sensitivity Score") +
      theme(legend.position = "none",
            plot.title = element_text(hjust = 0.5))+
      annotation_custom(grob_annot))
      
}

drugs.of.interest <- c("tamoxifen", "fulvestrant", "lapatinib", "MK-1775")

for(i in 1:length(drugs.of.interest)){
  PAM50_ISS_grapher(drugs.of.interest[i])
  Clin.sub_ISS_grapher(drugs.of.interest[i])
  TNBC_ISS_grapher(drugs.of.interest[i])
}


```

```{r volcano plot of Estimate vs p-value}

t.tst_preds.subtype <- t.tst_preds.subtype %>% 
  mutate(log10_p.value = -log10(bonf_p.value)) 



t.tst_preds.subtype %>% 
  filter(Drug %in% "MK-1775") %>% 
  ggplot(mapping = aes(x = estimate, y = log10_p.value, label = Drug)) + 
    geom_point(data = t.tst_preds.subtype) +
    geom_point(color = "red", size = 2.2) + 
    geom_text_repel(nudge_y = 5, nudge_x = -0.25) + 
    labs(title = "Volcano Plot of TNBC vs RPBC T-Test Results \n P-values and Estimate Scores for All Drugs", 
         y = "-log10(p.value)", 
         x = "Mean Differnce (RPBC - TNBC)", 
         caption = "Drugs More Effective in RPBC                                                                                                            Drugs More Effective in TNBC") + 
    theme(plot.title = element_text(hjust = 0.5))
    


```


```{r AOV Effect suze analysis and volcano plot}

#perform the effect size analysis to get eta-squared values
b_preds.subtype <- preds.subtype %>% 
  group_by(Drug) %>% 
  do(tidy(lsr::etaSquared(aov(.$PSS ~ .$PAM50)))) %>% 
  ungroup()

#we will also want to perform a Tukey HSD so that we can determine which group is the one that is most effective
HSD_preds.subtype <- preds.subtype %>% 
  group_by(Drug) %>% 
  do(tidy(TukeyHSD(aov(.$PSS ~ .$PAM50)))) %>% 
  ungroup()

#now joining the two datasets and for each drug determining which is the sensitive group
eta.sq_preds.subtype <- filter(b_preds.subtype) %>% 
  left_join(x= . , y = HSD_preds.subtype) %>% 
  group_by(Drug) %>% 
  summarise(sensitive_group = if_else(sum(estimate[1:4] > 0) == 4, 
                                      true = "basal", 
                                      false = if_else(sum(c(estimate[1] < 0,estimate[5:7] > 0)) == 4, 
                                                      true = "HER2", 
                                                      false = if_else(sum(estimate[c(2:3, 5:6)] < 0) == 4, 
                                                                      true = "luminal", 
                                                                      false = "none"))),
            eta.sq = eta.sq[1]) %>% 
  ungroup() %>% 
  left_join(., aov_preds.subtype) %>% 
  mutate(log10_p.value = -log10(bonf_p.value)) 

#print table (Supplemental table 1)
eta.sq_preds.subtype %>% 
  select(Drug, sensitive_group, eta.sq, bonf_p.value) %>% 
  arrange(bonf_p.value)


eta.sq_preds.subtype %>% 
  select(Drug, sensitive_group, eta.sq, bonf_p.value) %>% 
  group_by(sensitive_group) %>% 
  arrange(bonf_p.value) %>% 
  slice(1:5) 

# eta.sq_preds.subtype %>% 
#   filter(Drug %in% "MK-1775") %>% 
#   ggplot(mapping = aes(x = eta.sq, y = log10_p.value, label = Drug)) + 
#     geom_point(data = eta.sq_preds.subtype) +
#     geom_point(color = "red") + 
#     geom_text_repel(nudge_y = 5, nudge_x = -0.25) + 
#     labs(title = "Volcano Plot of TNBC vs RPBC ANOVA Results \n P-values and Eta Squared Values for All Drugs", 
#          y = "-log10(FDR)", 
#          x = "Eta Squared") + 
#     theme(plot.title = element_text(hjust = 0.5))
#     
# eta.sq_preds.subtype %>% 
#   ggplot(mapping = aes(x = log10_p.value, fill = sensitive_group)) + 
#     geom_histogram() 




```


## Correlations With Molecular Targets

We can look into this further by plotting Tamoxifen and Lapatinib ISS by ER and HER2 expression and seeing how well they correlate. 


```{r ER and HER2 correlations}
###########ER and Tamoxifen Response#################

#ER and Tamoxifen stats:

ER_t.tst_preds <- preds.subtype %>% 
  group_by(Drug) %>% 
  drop_na(er_status_by_ihc) %>% 
  do(tidy(t.test(PSS~er_status_by_ihc, data = .))) %>% 
  ungroup() %>% 
  mutate(bonf_p.value = p.adjust(p.value, method = "bonferroni")) %>% 
  arrange(bonf_p.value) %>% 
  filter(estimate > 0)

head(ER_t.tst_preds)

#ER positivity associates with Imputed Tamoxifen Response, it is the 5th most significant result
```

For ER, we see good correlation between ER status and imputed response to Tamoxifen. Additionally, we see that this imputed repsonse is associated with ER expression.


We can do a similar analysis for HER2. 

```{r}
############### HER2 and Lapatinib Response ##############################


HER2_t.tst_preds <- preds.subtype %>% 
  group_by(Drug) %>% 
  do(tidy(t.test(PSS~HER2_Subtype, data = .))) %>% 
  ungroup() %>% 
  mutate(bonf_p.value = p.adjust(p.value, method = "bonferroni")) %>% 
  arrange(bonf_p.value) %>% 
  filter(estimate > 0)

head(HER2_t.tst_preds)

# HER2 positiivity associates with Lapatinib response, is the 2nd most significant result. 

HER2_PSS <- preds.subtype %>% 
  filter(Drug == "lapatinib") %>% 
  drop_na(ERBB2_CN) %$%
  cor.test(PSS, ERBB2_CN, method = "spearman")

preds.subtype %>% 
  filter(Drug == "lapatinib") %>% 
  ggplot(aes(x = ERBB2_CN, y = PSS)) + 
    geom_point() +
    geom_smooth(method = "lm") + 
    annotate("text", x = 2, y = 14.1, label = paste("p-value =", HER2_PSS$p.value, ",  rho =", HER2_PSS$estimate, sep = "")) + 
    labs(title = "Imputed Lapatinib Response Correlates \n with HER2 Copy Number in TCGA BRCA Patients", 
         x = "ERBB2 Copy Number", 
         y = "Lapatinib Imputed Sensitivity Score (ISS)") + 
    theme(plot.title = element_text(hjust = 0.5))

```


For HER2, we see good correlation between ER status and imputed response to Tamoxifen. Additionally, we see that this imputed repsonse is associated with ER expression.



## TNBC Analysis 

We want to pick drugs that are better in TNBC by both effect size and p-value cutoffs. For this, we will use the top 20 most significant hits that have an effect size at least as great as taxomifen in ER+ cancers. Here is the list of drugs (with one drug combo) that fits this criteria


```{r AOV Beta analysis}
#perform the effect size analysis to get eta-squared values
b_preds.subtype <- preds.subtype %>% 
  group_by(Drug) %>% 
  do(tidy(lsr::etaSquared(aov(.$PSS ~ .$PAM50))))

#we will also want to perform a Tukey HSD so that we can determine which group is the one that is most effective
HSD_preds.subtype <- preds.subtype %>% 
  group_by(Drug) %>% 
  do(tidy(TukeyHSD(aov(.$PSS ~ .$PAM50)))) 

#now joining the two datasets and for each drug determining which is the sensitive group
eta.sq_preds.subtype <- filter(b_preds.subtype) %>% 
  left_join(x= . , y = HSD_preds.subtype) %>% 
  group_by(Drug) %>% 
  summarise(sensitive_group = if_else(sum(estimate[1:4] > 0) == 4, 
                                      true = "basal", 
                                      false = if_else(sum(c(estimate[1] < 0,estimate[5:7] > 0)) == 4, 
                                                      true = "HER2", 
                                                      false = if_else(sum(estimate[c(2:3, 5:6)] < 0) == 4, 
                                                                      true = "luminal", 
                                                                      false = "none"))),
            eta.sq = eta.sq[1]) %>% 
  left_join(., filter(aov_preds.subtype, term == ".$PAM50"))

eta.sq_preds.subtype %>% 
  filter(eta.sq > eta.sq[Drug == "tamoxifen"], 
         sensitive_group == "basal")


```


Now that we have a drug list, we can identify what pathways these drugs target. The list indicates that cell cycle is the most prominent pathway targeted. 

SAVE THE DRUG FILE TO USE IN ANALYZE IDWAS RESULTS!!!


```{r drug targets}

drug_targets <- read_csv("Drug_targets(review).csv") # From Ling and Gruener et al

targets_eta.sq <- eta.sq_preds.subtype %>% 
  filter(eta.sq > eta.sq[Drug == "tamoxifen"], 
         sensitive_group == "basal") %>% 
  select(Drug, sensitive_group, eta.sq, bonf_p.value) %>% 
  left_join(., drug_targets, by = c("Drug" = "Name in Database")) %>% 
  select(-c(7:9)) %>% 
  distinct() 

#Using the review information as well some manual curation, these are the simplified pathways for these compounds 
simplified_paths <- c("Amino Acid Metabolism", "Cell Cycle", "Cell Cycle", "Protein Stability and Degradation", "Cell Cycle", "Unknown", "Cell Cycle", "Sphingolipid Metabolism", "Cell Cycle", "Cell Cycle", "Protein Stability and Degradation", "STAT Signaling", "Cell Cycle", "Notch Signaling", "Purine Metabolism")

names(simplified_paths) <- c("968", "AZD7762", "BI-2536", "bortezomib", "BRD-K30748066", "BRD-K66453893", "BRD9876:MK-1775 (4:1 mol/mol)", "ceranib-2", "GSK461364", "MK-1775", "MLN2238", "NSC 74859", "PHA-793887", "semagacestat", "triazolothiadiazine")

targets_eta.sq <- targets_eta.sq %>% 
  mutate(Pathway = simplified_paths[names(simplified_paths) == Drug]) %>% 
  select(1:4, Pathway)

targets_eta.sq

```

We can go further and use the entire TCGA patient data and find associations between genetic features (i.e. non-synonymous mutations) and the patient's Imputed Sensitivty Scores. If we graph the 9 genes with the most significant values, this is the result.  



```{r}
t.tst_preds.subtype %>% mutate(Sensitive_Group = if_else(estimate > 0, "TNBC", "RPBC")) %>% 
  select(Drug, Sensitive_Group, bonf_p.value) %>%  arrange(bonf_p.value) 
```

